### ISSUE TITLE
Why do not put var in linked_dirs to shared

### ISSUE THREAD

**Issue by dooblem** [davidalger/capistrano-magento2#124]
Hello,
Today I got a problem when trying to generate varnish configuration from m2 backoffice (generated in var/varnish.vcl).

Is there a reason we are not putting the whole var directory in shared ?
I got only those things not in shared in my var: var/.htaccess and var/view_preprocessed and var/varnish.vcl.

Is view_preprocessed the only reason ? Is it better to keep it in the release ?

Looking on a magento cloud project, in magento cloud all var directory is mount binded in NFS. (only var/log is mount binded locally for performance reasons probably).

Thanks in advance,
Marc

**Comment by davidalger** [comment]
Hi Marc!

The contents of var is a tricky one. You are correct, the `view_preprocessed` directory and it's contents are the reason this entire dir is not shared among nodes. In older versions of Magento, the `./generated/` directory contents (generated via `setup:di:compile`) were also located in `var` adding to the mess.

With files in this directory being used at run-time (for `view_preprocessed` depends on magento mode, for older versions, the DI compiled files as well) the performance of Magento suffers dramatically because no NFS mount can hold a light to the performance of having files on a local disk. In Magento cloud, I believe they might GlusterFS to distribute things, which is perhaps better than NFS but has it's own unique scaling problems. Don't remember :)

> var/log is mount binded locally for performance reasons probably
That would be why, as writing to a log file on a shared mount would be worse than a file read.

So in short, the idea is share are little as possible. If you want to try it differently, you are definitely welcome to give it a shot.

My advice regarding the VCL, you can certainly use the admin to generate the file but I would copy it into `/etc/varnish/` on each of your hosts, not symlink to what's in the var dir.

**Comment by dooblem** [comment]
Thanks David.

The problem I got is a 50x error when configuring varnish in the BO. To avoid that can we add the var/varnish.vcl file in linked files ?
(or maybe just as a config or line to uncomment ?)

> With files in this directory being used at run-time (for `view_preprocessed` depends on magento mode, for older versions, the DI compiled files as well) the performance of Magento suffers dramatically because no NFS mount can hold a light to the performance of having files on a local disk. In Magento cloud, I believe they might GlusterFS to distribute things, which is perhaps better than NFS but has it's own unique scaling problems. Don't remember :)
> 
> > var/log is mount binded locally for performance reasons probably
> > That would be why, as writing to a log file on a shared mount would be worse than a file read.
> 
> So in short, the idea is share are little as possible. If you want to try it differently, you are definitely welcome to give it a shot.

Are you storing all the shared directory in NFS ? On our setups the capistrano shared directory is kept local.
We just add symlinks from the capistrano shared dir to the NFS mount, for each "shared" directory we wish to be shared by all web fronts.

I will may be give it a try, but like you I do not like more that the necessary to be in shared...

> My advice regarding the VCL, you can certainly use the admin to generate the file but I would copy it into `/etc/varnish/` on each of your hosts, not symlink to what's in the var dir.

Don't worry this is exactly what I did ;)

**Comment by dooblem** [comment]
I'm a bit confused...
It just got another problem setting up a m2 in load balancing. With the var root directory in the release, so read only.
It seems like magento is writing a var/.regenerate file when generating product thumbs...

I'm really close to convince myself of putting the whole var directory in NFS, with just a few symlinks to make log directories local.

I took a look again in m2cloud deploy process and found the following:
- on a live running website, deployed since several days, the var/view_preprocessed dir is totally empty
- on another of my websites deployed with capistrano, not yet in production, I wiped the content of the directory, and seen nothing wrong
- looking into ece-tools, seems they are disabling the view_preprocessed cache:
https://github.com/magento/ece-tools/blob/develop/dist/.magento.env.yaml#L85
https://github.com/magento/ece-tools/blob/develop/src/Process/Deploy/PreDeploy/CleanViewPreprocessed.php#L63

Is this directory needed anyway ?
